File: 5-2.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Map
  attr_accessor :map

  def initialize
    @size = 1
    @map = Numo::UInt8.zeros(@size, @size)
  end

  def expand(x, y)
    if x >= @size or y >= @size
      expand = [x - @size + 1, y - @size + 1].max
      @map = @map.append(Numo::UInt8.zeros(expand, @size), axis: 0)
      @map = @map.append(Numo::UInt8.zeros(@size + expand, expand), axis: 1)
      @size += expand
    end
  end

  def [](x, y)
    expand(x, y)
    @map[x, y]
  end

  def []=(x, y, val)
    expand(x, y)
    @map[x, y] = val
  end

  def mark(x1, y1, x2, y2)
    expand([x1, x2].max, [y1, y2].max)

    case
    when (x2 - x1).zero?
      dx = 0
    when (x2 - x1).positive?
      dx = 1
    when (x2 - x1).negative?
      dx = -1
    end

    case
    when (y2 - y1).zero?
      dy = 0
    when (y2 - y1).positive?
      dy = 1
    when (y2 - y1).negative?
      dy = -1
    end

    xstart = x1
    ystart = y1
    x1, x2 = [x1, x2].sort
    y1, y2 = [y1, y2].sort
    count = [x2 - x1, y2 - y1].max

    0.upto(count) { |d| self[xstart + dx * d, ystart + dy * d] += 1 }
  end

  def score
    Array(@map.flatten).count { |x| x > 1 }
  end

  def to_s
    "<#{self.class}:\nsize: #{@size}\n#{@map.transpose.inspect}\n>"
  end

  def inspect
    "<#{self.class}:\nsize: #{@size}\n#{@map.transpose.inspect}\n>"
  end
end

lines = File.read('5.input').lines.map(&:strip)
line_format = /(?<x1>[[:digit:]]+),(?<y1>[[:digit:]]+) -> (?<x2>[[:digit:]]+),(?<y2>[[:digit:]]+)/

map = Map.new

lines.each do |line|
  line_format.match(line) do |l|
    map.mark(l['x1'].to_i,
             l['y1'].to_i,
             l['x2'].to_i,
             l['y2'].to_i)
  end
end

print map.score, "\n"

File: 5-2.py
#!/usr/bin/env python

import numpy
import re

class Map:
    def __init__(self):
        self.size = 1
        self.map = numpy.zeros((self.size, self.size), dtype=numpy.uint)

    def expand(self, x, y):
        if x >= self.size or y >= self.size:
            expand = max([x - self.size + 1, y - self.size + 1])
            self.map = numpy.append(self.map, numpy.zeros((expand, self.size), dtype=numpy.uint), axis=0)
            self.map = numpy.append(self.map, numpy.zeros((self.size + expand, expand), dtype=numpy.uint), axis=1)
            self.size += expand

    def __getitem__(self, coords):
        x, y = coords
        self.expand(x, y)
        return self.map[x, y]

    def __setitem__(self, coords, value):
        x, y = coords
        self.expand(x, y)
        self.map[x, y] = value

    def mark(self, x1, y1, x2, y2):
        self.expand(max([x1, x2]), max([y1, y2]))

        if (x2 - x1) == 0:
            dx = 0
        elif (x2 - x1) > 0:
            dx = 1
        elif (x2 - x1) < 0:
            dx = -1

        if (y2 - y1) == 0:
            dy = 0
        elif (y2 - y1) > 0:
            dy = 1
        elif (y2 - y1) < 0:
            dy = -1

        xstart = x1
        ystart = y1

        xl = [x1, x2]
        xl.sort()
        x1, x2 = xl

        yl = [y1, y2]
        yl.sort()
        y1, y2 = yl

        count = max([x2 - x1, y2 - y1])

        for d in range(count + 1):
            self[xstart + dx * d, ystart + dy * d] += 1

    def score(self):
        return len([x for x in self.map.flatten().tolist() if x > 1])

    def __str__(self):
        return f"<{self.__class__.__name__}:\nsize: {self.size}\n{self.map.transpose()}\nscore: {self.score()}>"


lines = [line.strip() for line in open('5.input').readlines()]

line_format = re.compile("(?P<x1>[0-9]+),(?P<y1>[0-9]+) -> (?P<x2>[0-9]+),(?P<y2>[0-9]+)")

map = Map()

for line in lines:
    l = line_format.match(line)
    map.mark(int(l['x1']),
             int(l['y1']),
             int(l['x2']),
             int(l['y2']))

print(map.score())
